linodecli/baked/parsing.py

```diff
@@ -0,0 +1,184 @@
+"""
+This module contains logic related to string parsing and replacement.
+"""
+
+import functools
+import re
+from html import unescape
+from typing import List, Tuple
+
+# Sentence delimiter, split on a period followed by any type of
+# whitespace (space, new line, tab, etc.)
+REGEX_SENTENCE_DELIMITER = re.compile(r"\.(?:\s|$)")
+
+# Matches on pattern __prefix__ at the beginning of a description
+# or after a comma
+REGEX_TECHDOCS_PREFIX = re.compile(r"(?:, |\A)__([\w-]+)__")
+
+# Matches on pattern [link title](https://.../)
+REGEX_MARKDOWN_LINK = re.compile(r"\[(?P<text>.*?)]\((?P<link>.*?)\)")
+
+MARKDOWN_RICH_TRANSLATION = [
+    # Inline code blocks (e.g. `cool code`)
+    (
+        re.compile(
+            r"`(?P<text>[^`]+)`",
+        ),
+        "italic deep_pink3 on grey15",
+    ),
+    # Bold tag (e.g. `**bold**` or `__bold__`)
+    (
+        re.compile(
+            r"\*\*(?P<text>[^_\s]+)\*\*",
+        ),
+        "b",
+    ),
+    (
+        re.compile(
+            r"__(?P<text>[^_\s]+)__",
+        ),
+        "b",
+    ),
+    # Italics tag (e.g. `*italics*` or `_italics_`)
+    (
+        re.compile(
+            r"\*(?P<text>[^*\s]+)\*",
+        ),
+        "i",
+    ),
+    (
+        re.compile(
+            r"_(?P<text>[^_\s]+)_",
+        ),
+        "i",
+    ),
+]
+
+
+def markdown_to_rich_markup(markdown: str) -> str:
+    """
+    This function returns a version of the given argument description
+    with the appropriate color tags.
+
+    NOTE, Rich does support Markdown rendering, but it isn't suitable for this
+    use-case quite yet due to some Group(...) padding issues and limitations
+    with syntax themes.
+
+    :param markdown: The argument description to colorize.
+    :type markdown: str
+
+    :returns: The translated Markdown
+    """
+
+    result = markdown
+
+    for exp, style in MARKDOWN_RICH_TRANSLATION:
+        result = exp.sub(
+            # Necessary to avoid cell-var-in-loop linter fer
+            functools.partial(
+                lambda style, match: f"[{style}]{match['text']}[/]", style
+            ),
+            result,
+        )
+
+    return result
+
+
+def extract_markdown_links(description: str) -> Tuple[str, List[str]]:
+    """
+    Extracts all Markdown links from the given description and
+    returns them alongside the stripped description.
+
+    :param description: The description of a CLI argument.
+    :type description: str
+
+    :returns: The stripped description and a list of extracted links.
+    :rtype: Tuple[str, List[str]]
+    """
+    result_links = []
+
+    def _sub_handler(match: re.Match) -> str:
+        link = match["link"]
+        if link.startswith("/"):
+            link = f"https://linode.com{link}"
```

Comment:
The TechDocs spec does _not_ use relative paths so this is necessary to maintain for backwards compatibility purposes